library/std/src/sys/uefi/stdio.rs

```diff
@@ -4,51 +4,72 @@ use crate::mem::MaybeUninit;
 use crate::os::uefi;
 use crate::ptr::NonNull;
 
-const MAX_BUFFER_SIZE: usize = 8192;
+pub struct Stdin {
+    pending: [Option<u8>; 2],
+}
 
-pub struct Stdin;
 pub struct Stdout;
 pub struct Stderr;
 
 impl Stdin {
     pub const fn new() -> Stdin {
-        Stdin
+        Stdin { pending: [None; 2] }
     }
 }
 
 impl io::Read for Stdin {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if buf.is_empty() {
+            return Ok(0);
+        }
+
         let st: NonNull<r_efi::efi::SystemTable> = uefi::env::system_table().cast();
         let stdin = unsafe { (*st.as_ptr()).con_in };
+        let mut count: usize = 0;
 
-        // Try reading any pending data
-        let inp = match read_key_stroke(stdin) {
-            Ok(x) => x,
-            Err(e) if e == r_efi::efi::Status::NOT_READY => {
-                // Wait for keypress for new data
-                wait_stdin(stdin)?;
-                read_key_stroke(stdin).map_err(|x| io::Error::from_raw_os_error(x.as_usize()))?
-            }
-            Err(e) => {
-                return Err(io::Error::from_raw_os_error(e.as_usize()));
+        // Write any pending character
+        while let Some(byte) = self.pending[0] {
+            buf[count] = byte;
+            count += 1;
+            self.pending[0] = self.pending[1].take();
+
+            if count >= buf.len() {
+                return Ok(count);
             }
-        };
+        }
 
-        // Check if the key is printiable character
-        if inp.scan_code != 0x00 {
+        // Try reading new character
+        let inp = simple_text_input_read(stdin)?;
+        if inp == 0x00 {
             return Err(io::const_io_error!(io::ErrorKind::Interrupted, "Special Key Press"));
         }
 
-        // SAFETY: Iterator will have only 1 character since we are reading only 1 Key
-        // SAFETY: This character will always be UCS-2 and thus no surrogates.
-        let ch: char = char::decode_utf16([inp.unicode_char]).next().unwrap().unwrap();
-        if ch.len_utf8() > buf.len() {
-            return Ok(0);
+        let Some(Ok(ch)) = char::decode_utf16([inp]).next() else {
+            return Err(io::const_io_error!(io::ErrorKind::InvalidInput, "Invalid Input"));
+        };
+
+        // This will always be > 0
+        let buf_free_count = buf.len() - count;
+        assert!(buf_free_count > 0);
+
+        if ch.len_utf8() > buf_free_count {
+            let mut temp = [0u8; 3];
```

Comment:
3? It can take 4 bytes, no? Or is this relying on the input being UCS-2 encoding instead of UTF-16?

Reply:
It is relying on the input being UCS-2 (aka no surrogates). The spec actually disallows surrogates, so it should be a safe assumption to make for the standard UEFI protocols.

Reply:
From the platform support docs:

> While the strings in UEFI should be valid UCS-2, in practice, many implementations just do not care and use UTF-16 strings.
> Thus, the current implementation supports full UTF-16 strings.

So I would prefer that we handle the full 4 byte case in this code, but give ourselves permission to emit errors or panic or whatever we want if we encounter something that clearly came from a surrogate pair. This will let our code be more resilient against actual UEFI implementations, and also make it easier to change our minds in the future.

Reply:
This also means we can reuse more of the code from `std::sys::windows::stdio`.